// @vuetify/vite-plugin/dist/importPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.importPlugin = void 0;
const path_1 = require("path");
const loader_shared_1 = require("@vuetify/loader-shared");
const url_1 = require("url");
function parseId(id) {
    const { query, pathname } = (0, url_1.parse)(id);
    return {
        query: query ? Object.fromEntries(new url_1.URLSearchParams(query)) : null,
        path: pathname !== null && pathname !== void 0 ? pathname : id
    };
}
function importPlugin() {
    return {
        name: 'vuetify:import',
        async transform(code, id) {
            const { query, path } = parseId(id);
            if ((0, path_1.extname)(path) === '.vue' && !query) {
                if (/^import { render as _sfc_render } from ".*"$/m.test(code)) {
                    return null;
                }
                const { code: imports, source } = (0, loader_shared_1.generateImports)(code);
                return source + imports;
            }
            else if (query && 'vue' in query && query.type === 'template') {
                const { code: imports, source } = (0, loader_shared_1.generateImports)(code);
                return source + imports;
            }
            return null;
        }
    };
}
exports.importPlugin = importPlugin;
